File: scripts/list-component-classes.py
#!/usr/bin/env python3
"""
list-component-classes.py — Gram component class catalog generator

Reads all installed Gram plugin class files and prints every available
component class (id, name, icon path, source plugin).  Useful for AI
agents and developers when assigning tech-stack classes to components
in a threat-model JSON.

Usage (run from repo root):
    python3 scripts/list-component-classes.py
    python3 scripts/list-component-classes.py --format json
    python3 scripts/list-component-classes.py --filter aws
"""

import argparse
import json
import re
import sys
from pathlib import Path


PLUGIN_CLASS_FILES = {
    "klarna":    "plugins/klarna/src/classes.ts",
    "aws":       "plugins/aws/src/classes.ts",
    "azure":     "plugins/azure/src/classes.ts",
    "svgporn":   "plugins/svgporn/src/classes.ts",
    "kubernetes":"plugins/kubernetes/src/classes.ts",
    "cncf":      "plugins/cncf/src/classes.ts",
}


def parse_ts_classes(filepath: Path) -> list[dict]:
    """Parse { id, name, icon, componentType } object literals from a TS file."""
    content = filepath.read_text()
    entries = []
    for block in re.findall(r'\{[^{}]+\}', content, re.DOTALL):
        id_m   = re.search(r'id:\s*["\']([^"\']+)["\']', block)
        name_m = re.search(r'name:\s*["\']([^"\']+)["\']', block)
        icon_m = re.search(r'icon:\s*["\']([^"\']+)["\']', block)
        type_m = re.search(r'componentType:\s*["\']([^"\']+)["\']', block)
        if id_m and name_m:
            entries.append({
                "id":            id_m.group(1),
                "name":          name_m.group(1),
                "icon":          icon_m.group(1) if icon_m else "",
                "componentType": type_m.group(1) if type_m else "any",
            })
    return entries


def main():
    parser = argparse.ArgumentParser(description=__doc__,
                                     formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("--format", choices=["table", "json"], default="table",
                        help="Output format (default: table)")
    parser.add_argument("--filter", metavar="TERM",
                        help="Case-insensitive substring filter on class name")
    parser.add_argument("--plugin", choices=list(PLUGIN_CLASS_FILES.keys()),
                        help="Restrict to a single plugin")
    args = parser.parse_args()

    repo_root = Path(__file__).parent.parent
    all_classes = []

    plugins = {args.plugin: PLUGIN_CLASS_FILES[args.plugin]} if args.plugin else PLUGIN_CLASS_FILES
    for plugin_name, rel_path in plugins.items():
        fpath = repo_root / rel_path
        if not fpath.exists():
            print(f"WARNING: {fpath} not found — skipping", file=sys.stderr)
            continue
        for entry in parse_ts_classes(fpath):
            all_classes.append({**entry, "plugin": plugin_name})

    if args.filter:
        term = args.filter.lower()
        all_classes = [c for c in all_classes if term in c["name"].lower()]

    all_classes.sort(key=lambda c: (c["plugin"], c["name"].lower()))

    if args.format == "json":
        print(json.dumps(all_classes, indent=2))
        return

    # Table output
    col_w = {"plugin": 10, "name": 50, "id": 36, "icon": 70}
    header = (f"{'PLUGIN':<{col_w['plugin']}}  {'NAME':<{col_w['name']}}  "
              f"{'ID':<{col_w['id']}}  ICON")
    sep = "-" * (len(header) + 10)
    print(sep)
    print(header)
    print(sep)
    for c in all_classes:
        print(f"{c['plugin']:<{col_w['plugin']}}  "
              f"{c['name']:<{col_w['name']}}  "
              f"{c['id']:<{col_w['id']}}  "
              f"{c['icon']}")
    print(sep)
    print(f"Total: {len(all_classes)} classes")


if __name__ == "__main__":
    main()

File: scripts/list-component-classes.mjs
#!/usr/bin/env node
/**
 * list-component-classes.mjs — Gram component class catalog generator
 *
 * Reads all installed Gram plugin class files and prints every available
 * component class (id, name, icon path, source plugin). Useful for AI
 * agents and developers when assigning tech-stack classes to components
 * in a threat-model JSON.
 *
 * Usage (run from repo root):
 *   node scripts/list-component-classes.mjs
 *   node scripts/list-component-classes.mjs --format json
 *   node scripts/list-component-classes.mjs --filter aws
 *   node scripts/list-component-classes.mjs --plugin aws
 */

import { existsSync, readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";

const PLUGIN_CLASS_FILES = {
  aws: "plugins/aws/src/classes.ts",
  azure: "plugins/azure/src/classes.ts",
  svgporn: "plugins/svgporn/src/classes.ts",
  kubernetes: "plugins/kubernetes/src/classes.ts",
  cncf: "plugins/cncf/src/classes.ts",
};

function parseTsClasses(filepath) {
  const content = readFileSync(filepath, "utf8");
  const blocks = content.match(/\{[^{}]+\}/g) ?? [];
  const entries = [];
  for (const block of blocks) {
    const idMatch = block.match(/id:\s*["']([^"']+)["']/);
    const nameMatch = block.match(/name:\s*["']([^"']+)["']/);
    const iconMatch = block.match(/icon:\s*["']([^"']+)["']/);
    const typeMatch = block.match(/componentType:\s*["']([^"']+)["']/);
    if (idMatch && nameMatch) {
      entries.push({
        id: idMatch[1],
        name: nameMatch[1],
        icon: iconMatch ? iconMatch[1] : "",
        componentType: typeMatch ? typeMatch[1] : "any",
      });
    }
  }
  return entries;
}

function printHelp() {
  process.stdout.write(
    [
      "Usage: node scripts/list-component-classes.mjs [options]",
      "",
      "Options:",
      "  --format <table|json>   Output format (default: table)",
      "  --filter <term>         Case-insensitive substring filter on class name",
      `  --plugin <name>         Restrict to a single plugin (${Object.keys(PLUGIN_CLASS_FILES).join("|")})`,
      "  --help                  Show this help and exit",
      "",
    ].join("\n")
  );
}

function padRight(str, width) {
  const s = String(str);
  return s.length >= width ? s : s + " ".repeat(width - s.length);
}

function main() {
  const { values } = parseArgs({
    options: {
      format: { type: "string", default: "table" },
      filter: { type: "string" },
      plugin: { type: "string" },
      help: { type: "boolean", default: false },
    },
    strict: true,
    allowPositionals: false,
  });

  if (values.help) {
    printHelp();
    return;
  }

  if (!["table", "json"].includes(values.format)) {
    process.stderr.write(`Invalid --format: ${values.format} (use table|json)\n`);
    process.exit(2);
  }

  if (values.plugin && !(values.plugin in PLUGIN_CLASS_FILES)) {
    process.stderr.write(
      `Unknown --plugin: ${values.plugin} (use ${Object.keys(PLUGIN_CLASS_FILES).join("|")})\n`
    );
    process.exit(2);
  }

  const repoRoot = resolve(dirname(fileURLToPath(import.meta.url)), "..");
  const plugins = values.plugin
    ? { [values.plugin]: PLUGIN_CLASS_FILES[values.plugin] }
    : PLUGIN_CLASS_FILES;

  let allClasses = [];
  for (const [pluginName, relPath] of Object.entries(plugins)) {
    const fpath = resolve(repoRoot, relPath);
    if (!existsSync(fpath)) {
      process.stderr.write(`WARNING: ${fpath} not found — skipping\n`);
      continue;
    }
    for (const entry of parseTsClasses(fpath)) {
      allClasses.push({ ...entry, plugin: pluginName });
    }
  }

  if (values.filter) {
    const term = values.filter.toLowerCase();
    allClasses = allClasses.filter((c) => c.name.toLowerCase().includes(term));
  }

  allClasses.sort((a, b) => {
    if (a.plugin !== b.plugin) return a.plugin < b.plugin ? -1 : 1;
    const an = a.name.toLowerCase();
    const bn = b.name.toLowerCase();
    if (an !== bn) return an < bn ? -1 : 1;
    return 0;
  });

  if (values.format === "json") {
    process.stdout.write(JSON.stringify(allClasses, null, 2) + "\n");
    return;
  }

  const colW = { plugin: 10, name: 50, id: 36, icon: 70 };
  const header =
    `${padRight("PLUGIN", colW.plugin)}  ` +
    `${padRight("NAME", colW.name)}  ` +
    `${padRight("ID", colW.id)}  ` +
    "ICON";
  const sep = "-".repeat(header.length + 10);

  process.stdout.write(sep + "\n");
  process.stdout.write(header + "\n");
  process.stdout.write(sep + "\n");
  for (const c of allClasses) {
    process.stdout.write(
      `${padRight(c.plugin, colW.plugin)}  ` +
        `${padRight(c.name, colW.name)}  ` +
        `${padRight(c.id, colW.id)}  ` +
        `${c.icon}\n`
    );
  }
  process.stdout.write(sep + "\n");
  process.stdout.write(`Total: ${allClasses.length} classes\n`);
}

main();
